src/FreeSpaceEllipse.cpp

```diff
@@ -0,0 +1,142 @@
+    assert(ok);
+}
+
+bool FreeSpaceEllipse::isPointInside(const iDynTree::VectorFixSize<2> &testPoint) const
+{
+    if (!m_isSet)
+    {
+        return true; //Default case
+    }
+
+    return generatorsModule(computeGenerators(testPoint)) <= 1.0;
+}
+
+bool FreeSpaceEllipse::setEllipse(double a, double b, double theta, double centerOffsetX, double centerOffsetY)
+{
+    if (a < 1e-10 || b < 1e-10)
+    {
+        std::cerr << "[FreeSpaceEllipse::setEllipse] The input a and b are supposed to be non-zero." << std::endl;
+        return false;
+    }
+
+    Eigen::Matrix2d rotationMatrix;
+    rotationMatrix(0,0) = std::cos(theta);
+    rotationMatrix(0,1) = -std::sin(theta);
+    rotationMatrix(1,0) = std::sin(theta);
+    rotationMatrix(1,1) =std::cos(theta);
+
+    Eigen::Matrix2d defaultC;
+    defaultC.setZero();
+    defaultC(0,0) = a;
+    defaultC(1,1) = b;
+
+    iDynTree::toEigen(m_C) = rotationMatrix * defaultC;
+    iDynTree::toEigen(m_C_inverse) = iDynTree::toEigen(m_C).inverse();
+    m_d(0) = centerOffsetX;
+    m_d(1) = centerOffsetY;
+    m_isSet = true;
+
+    return true;
+}
+
+const iDynTree::VectorFixSize<2> &FreeSpaceEllipse::centerOffset() const
+{
+    return m_d;
+}
+
+const iDynTree::MatrixFixSize<2, 2> &FreeSpaceEllipse::imageMatrix() const
+{
+    return m_C;
+}
+
+bool FreeSpaceEllipse::setEllipse(const iDynTree::MatrixFixSize<2, 2> &imageMatrix, const iDynTree::VectorFixSize<2> &centerOffset)
+{
+    if (iDynTree::toEigen(imageMatrix).determinant() < 1e-10)
+    {
+        std::cerr << "[FreeSpaceEllipse::setEllipse] The input image matrix is almost singular." << std::endl;
+        return false;
+    }
+
+    m_C = imageMatrix;
+    iDynTree::toEigen(m_C_inverse) = iDynTree::toEigen(m_C).inverse();
+    m_d = centerOffset;
+    m_isSet = true;
+
+    return true;
+}
+
+iDynTree::Vector2 FreeSpaceEllipse::computeGenerators(const iDynTree::VectorFixSize<2> &inputPoint) const
+{
+    iDynTree::Vector2 output;
+    iDynTree::toEigen(output) = iDynTree::toEigen(m_C_inverse) * (iDynTree::toEigen(inputPoint) - iDynTree::toEigen(m_d));
+    return output;
+}
+
+double FreeSpaceEllipse::generatorsModule(const iDynTree::VectorFixSize<2> &generators) const
+{
+    return (iDynTree::toEigen(generators)).norm();
+}
+
+FreeSpaceEllipse::FreeSpaceEllipse()
+{
+    iDynTree::toEigen(m_C).setIdentity();
+    m_C_inverse = m_C;
+    m_d.zero();
+    m_isSet = false;
+}
+
+iDynTree::Vector2 FreeSpaceEllipse::projectPointInsideEllipse(const iDynTree::VectorFixSize<2> &testPoint) const
+{
+    if (!m_isSet)
+    {
+        return testPoint; //Default case
+    }
+
+    iDynTree::Vector2 generators = computeGenerators(testPoint);
+    double module = generatorsModule(generators);
+    if (module <= 1.0)
+    {
+        return testPoint;
+    }
+
+    iDynTree::VectorFixSize<2> normalizedGenerators;
+
+    iDynTree::toEigen(normalizedGenerators) = iDynTree::toEigen(generators) / module;
```

Comment:
Check also for very small values of `module` to be careful about divide-by-zero situations.

Reply:
In this case we are sure that module > 1.0 because of https://github.com/robotology/unicycle-footstep-planner/pull/40/files/76be3a8e8c361012f7e32d2f9a43f1c6b5e468d0#diff-c78d6c59bdc612f9a1e6a91a96be420ecc01b8b979a9cd3f12baf4850fe0b319R120-R123

Reply:
Oh yeah, sorry! Should've thought twice :D